File: packages/not-vue/bundle.py
from __future__ import annotations

import argparse
import atexit
import http.server
import shutil
import signal
import socketserver
import subprocess
import sys
import tempfile
import threading
from dataclasses import dataclass
from pathlib import Path
from typing import Callable


DEFAULT_PORT = 3000
TEMPLATE_PLACEHOLDER = "<!-- TEMPLATE -->"


@dataclass(frozen=True)
class BundleConfig:
    root_dir: Path
    template_html_path: Path
    index_html_path: Path
    vendor_ts_path: Path
    index_ts_path: Path
    vendor_js_path: Path
    index_js_path: Path


def new_bundle_config() -> BundleConfig:
    root_dir = Path(__file__).resolve().parent
    return BundleConfig(
        root_dir=root_dir,
        template_html_path=root_dir / "template.html",
        index_html_path=root_dir / "index.html",
        vendor_ts_path=root_dir / "vendor.ts",
        index_ts_path=root_dir / "index.ts",
        vendor_js_path=root_dir / "vendor.js",
        index_js_path=root_dir / "index.js",
    )


def run(mode: str) -> int:
    config = new_bundle_config()
    if mode == "dev":
        return run_dev(config)
    if mode == "build":
        return run_build(config)
    raise ValueError(f'unknown mode "{mode}", expected one of: dev, build')


def run_build(config: BundleConfig) -> int:
    cleanup = compile_typescript(config, remove_output_on_cleanup=False)
    try:
        print(f"build complete: {config.index_js_path}, {config.vendor_js_path}")
        return 0
    finally:
        cleanup()


def run_dev(config: BundleConfig) -> int:
    cleanup = compile_typescript(config, remove_output_on_cleanup=True)
    atexit.register(cleanup)

    class RequestHandler(http.server.SimpleHTTPRequestHandler):
        def __init__(self, *args, **kwargs):
            super().__init__(*args, directory=str(config.root_dir), **kwargs)

        def do_GET(self) -> None:
            if self.path == "/":
                self.path = "/index.html"
            return super().do_GET()

    class ThreadingTCPServer(socketserver.ThreadingTCPServer):
        allow_reuse_address = True

    server = ThreadingTCPServer(("", DEFAULT_PORT), RequestHandler)
    stop_event = threading.Event()

    def shutdown_handler(_signum: int, _frame) -> None:
        if stop_event.is_set():
            return
        stop_event.set()
        print("shutting down dev server...")
        threading.Thread(target=server.shutdown, daemon=True).start()

    previous_sigint = signal.getsignal(signal.SIGINT)
    previous_sigterm = signal.getsignal(signal.SIGTERM)
    signal.signal(signal.SIGINT, shutdown_handler)
    signal.signal(signal.SIGTERM, shutdown_handler)

    try:
        print(f"dev server running at http://localhost:{DEFAULT_PORT}")
        server.serve_forever()
        return 0
    finally:
        server.server_close()
        cleanup()
        signal.signal(signal.SIGINT, previous_sigint)
        signal.signal(signal.SIGTERM, previous_sigterm)


def compile_typescript(
    config: BundleConfig,
    remove_output_on_cleanup: bool,
) -> Callable[[], None]:
    generated_entry_path, cleanup_temp = prepare_generated_sources(config)
    cleanup_fns: list[Callable[[], None]] = [cleanup_temp]

    temp_output_dir = Path(tempfile.mkdtemp(prefix="proxy-tsc-out-"))
    cleanup_fns.append(lambda: shutil.rmtree(temp_output_dir, ignore_errors=True))

    try:
        run_tsc(generated_entry_path, temp_output_dir)
        compiled_index_js_path = temp_output_dir / "index.js"
        compiled_vendor_js_path = temp_output_dir / "vendor.js"

        if not compiled_index_js_path.exists():
            raise FileNotFoundError(
                f"compiled output missing: {compiled_index_js_path}"
            )
        if not compiled_vendor_js_path.exists():
            raise FileNotFoundError(
                f"compiled output missing: {compiled_vendor_js_path}"
            )

        copy_file(compiled_index_js_path, config.index_js_path)
        copy_file(compiled_vendor_js_path, config.vendor_js_path)

        if remove_output_on_cleanup:
            cleanup_fns.append(lambda: safe_unlink(config.index_js_path))
            cleanup_fns.append(lambda: safe_unlink(config.vendor_js_path))
    except Exception:
        run_cleanup(cleanup_fns)
        raise

    def cleanup() -> None:
        run_cleanup(cleanup_fns)

    return cleanup


def prepare_generated_sources(config: BundleConfig) -> tuple[Path, Callable[[], None]]:
    template_content = config.template_html_path.read_text(encoding="utf-8")
    vendor_ts_content = config.vendor_ts_path.read_text(encoding="utf-8")
    index_ts_content = config.index_ts_path.read_text(encoding="utf-8")

    if TEMPLATE_PLACEHOLDER not in index_ts_content:
        raise ValueError(f'placeholder "{TEMPLATE_PLACEHOLDER}" not found in index.ts')

    replaced = index_ts_content.replace(TEMPLATE_PLACEHOLDER, template_content, 1)

    temp_dir = Path(tempfile.mkdtemp(prefix="proxy-ts-src-"))
    generated_index_path = temp_dir / "index.ts"
    generated_vendor_path = temp_dir / "vendor.ts"

    generated_index_path.write_text(replaced, encoding="utf-8")
    generated_vendor_path.write_text(vendor_ts_content, encoding="utf-8")

    def cleanup() -> None:
        shutil.rmtree(temp_dir, ignore_errors=True)

    return generated_index_path, cleanup


def run_tsc(entry_path: Path, out_dir: Path) -> None:
    command = resolve_tsc_command(entry_path, out_dir)
    try:
        subprocess.run(command, check=True)
    except subprocess.CalledProcessError as error:
        raise RuntimeError(f"run typescript compiler: {error}") from error


def resolve_tsc_command(entry_path: Path, out_dir: Path) -> list[str]:
    base_args = [
        "--target",
        "ESNext",
        "--module",
        "ESNext",
        "--moduleResolution",
        "Bundler",
        "--lib",
        "ESNext,DOM",
        "--strict",
        "--skipLibCheck",
        "--pretty",
        "false",
        "--declaration",
        "false",
        "--sourceMap",
        "false",
        "--removeComments",
        "false",
        "--outDir",
        str(out_dir),
        str(entry_path),
    ]

    if shutil.which("tsc"):
        return ["tsc", *base_args]
    return ["npx", "--yes", "tsc", *base_args]


def copy_file(src_path: Path, dst_path: Path) -> None:
    shutil.copyfile(src_path, dst_path)


def safe_unlink(path: Path) -> None:
    try:
        path.unlink()
    except FileNotFoundError:
        pass


def run_cleanup(cleanups: list[Callable[[], None]]) -> None:
    for cleanup in reversed(cleanups):
        try:
            cleanup()
        except Exception:
            pass


def parse_args(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="mini-vue build/dev tool")
    parser.add_argument(
        "mode",
        nargs="?",
        default="dev",
        choices=("dev", "build"),
        help="run mode",
    )
    return parser.parse_args(argv)


def main(argv: list[str] | None = None) -> int:
    args = parse_args(argv or sys.argv[1:])
    try:
        return run(args.mode)
    except Exception as error:
        print(error, file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/not-vue/bundle.go
package main

import (
	"context"
	"errors"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"os/exec"
	"os/signal"
	"path/filepath"
	"runtime"
	"strings"
	"syscall"
	"time"
)

const (
	defaultPort         = "3000"
	templatePlaceholder = "<!-- TEMPLATE -->"
)

type bundleConfig struct {
	rootDir          string
	templateHTMLPath string
	indexHTMLPath    string
	vendorTSPath     string
	indexTSPath      string
	vendorJSPath     string
	indexJSPath      string
}

func main() {
	log.SetFlags(0)

	if err := run(os.Args[1:]); err != nil {
		log.Fatal(err)
	}
}

func run(args []string) error {
	mode := "dev"
	if len(args) > 0 {
		mode = args[0]
	}

	config, err := newBundleConfig()
	if err != nil {
		return err
	}

	switch mode {
	case "dev":
		return runDev(config)
	case "build":
		return runBuild(config)
	default:
		return fmt.Errorf("unknown mode %q, expected one of: dev, build", mode)
	}
}

func newBundleConfig() (bundleConfig, error) {
	_, filePath, _, ok := runtime.Caller(0)
	if !ok {
		return bundleConfig{}, errors.New("resolve bundle.go location: runtime caller unavailable")
	}
	rootDir := filepath.Dir(filePath)

	return bundleConfig{
		rootDir:          rootDir,
		templateHTMLPath: filepath.Join(rootDir, "template.html"),
		indexHTMLPath:    filepath.Join(rootDir, "index.html"),
		vendorTSPath:     filepath.Join(rootDir, "vendor.ts"),
		indexTSPath:      filepath.Join(rootDir, "index.ts"),
		vendorJSPath:     filepath.Join(rootDir, "vendor.js"),
		indexJSPath:      filepath.Join(rootDir, "index.js"),
	}, nil
}

func runBuild(config bundleConfig) error {
	cleanup, err := compileTypeScript(config, false)
	if err != nil {
		return err
	}
	defer cleanup()

	log.Printf("build complete: %s, %s", config.indexJSPath, config.vendorJSPath)
	return nil
}

func runDev(config bundleConfig) error {
	cleanup, err := compileTypeScript(config, true)
	if err != nil {
		return err
	}
	defer cleanup()

	mux := http.NewServeMux()
	fileServer := http.FileServer(http.Dir(config.rootDir))
	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path == "/" {
			http.ServeFile(w, r, config.indexHTMLPath)
			return
		}
		fileServer.ServeHTTP(w, r)
	})

	server := &http.Server{
		Addr:              ":" + defaultPort,
		Handler:           mux,
		ReadHeaderTimeout: 5 * time.Second,
	}

	errCh := make(chan error, 1)
	go func() {
		log.Printf("dev server running at http://localhost:%s", defaultPort)
		if serveErr := server.ListenAndServe(); serveErr != nil && !errors.Is(serveErr, http.ErrServerClosed) {
			errCh <- serveErr
		}
		close(errCh)
	}()

	signalCtx, stop := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
	defer stop()

	select {
	case <-signalCtx.Done():
		log.Println("shutting down dev server...")
	case serveErr := <-errCh:
		if serveErr != nil {
			return fmt.Errorf("start dev server: %w", serveErr)
		}
		return nil
	}

	shutdownCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	if err := server.Shutdown(shutdownCtx); err != nil {
		return fmt.Errorf("shutdown dev server: %w", err)
	}

	return nil
}

func compileTypeScript(config bundleConfig, removeOutputOnCleanup bool) (func(), error) {
	generatedEntryPath, cleanupTemp, err := prepareGeneratedSources(config)
	if err != nil {
		return nil, err
	}

	var cleanupFns []func()
	cleanupFns = append(cleanupFns, cleanupTemp)

	tempOutputDir, err := os.MkdirTemp("", "proxy-tsc-out-*")
	if err != nil {
		runCleanup(cleanupFns)
		return nil, fmt.Errorf("create temp output dir: %w", err)
	}
	cleanupFns = append(cleanupFns, func() {
		_ = os.RemoveAll(tempOutputDir)
	})

	if err := runTSC(generatedEntryPath, tempOutputDir); err != nil {
		runCleanup(cleanupFns)
		return nil, err
	}

	compiledIndexJSPath := filepath.Join(tempOutputDir, "index.js")
	if _, err := os.Stat(compiledIndexJSPath); err != nil {
		runCleanup(cleanupFns)
		return nil, fmt.Errorf("compiled output missing: %w", err)
	}

	compiledVendorJSPath := filepath.Join(tempOutputDir, "vendor.js")
	if _, err := os.Stat(compiledVendorJSPath); err != nil {
		runCleanup(cleanupFns)
		return nil, fmt.Errorf("compiled output missing: %w", err)
	}

	if err := copyFile(compiledIndexJSPath, config.indexJSPath); err != nil {
		runCleanup(cleanupFns)
		return nil, fmt.Errorf("write %s: %w", config.indexJSPath, err)
	}

	if err := copyFile(compiledVendorJSPath, config.vendorJSPath); err != nil {
		runCleanup(cleanupFns)
		return nil, fmt.Errorf("write %s: %w", config.vendorJSPath, err)
	}

	if removeOutputOnCleanup {
		cleanupFns = append(cleanupFns, func() {
			_ = os.Remove(config.indexJSPath)
			_ = os.Remove(config.vendorJSPath)
		})
	}

	return func() {
		runCleanup(cleanupFns)
	}, nil
}

func prepareGeneratedSources(config bundleConfig) (string, func(), error) {
	templateContent, err := os.ReadFile(config.templateHTMLPath)
	if err != nil {
		return "", nil, fmt.Errorf("read template.html: %w", err)
	}

	vendorTSContent, err := os.ReadFile(config.vendorTSPath)
	if err != nil {
		return "", nil, fmt.Errorf("read vendor.ts: %w", err)
	}

	indexTSContent, err := os.ReadFile(config.indexTSPath)
	if err != nil {
		return "", nil, fmt.Errorf("read index.ts: %w", err)
	}

	indexTSString := string(indexTSContent)
	if !strings.Contains(indexTSString, templatePlaceholder) {
		return "", nil, fmt.Errorf("placeholder %q not found in index.ts", templatePlaceholder)
	}

	replaced := strings.Replace(indexTSString, templatePlaceholder, string(templateContent), 1)

	tempDir, err := os.MkdirTemp("", "proxy-ts-src-*")
	if err != nil {
		return "", nil, fmt.Errorf("create temp source dir: %w", err)
	}

	generatedIndexPath := filepath.Join(tempDir, "index.ts")
	if err := os.WriteFile(generatedIndexPath, []byte(replaced), 0o644); err != nil {
		_ = os.RemoveAll(tempDir)
		return "", nil, fmt.Errorf("write generated index.ts: %w", err)
	}

	generatedVendorPath := filepath.Join(tempDir, "vendor.ts")
	if err := os.WriteFile(generatedVendorPath, vendorTSContent, 0o644); err != nil {
		_ = os.RemoveAll(tempDir)
		return "", nil, fmt.Errorf("write generated vendor.ts: %w", err)
	}

	return generatedIndexPath, func() {
		_ = os.RemoveAll(tempDir)
	}, nil
}

func runTSC(entryPath, outDir string) error {
	command, args := resolveTSCCommand(entryPath, outDir)

	cmd := exec.Command(command, args...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	if err := cmd.Run(); err != nil {
		return fmt.Errorf("run typescript compiler: %w", err)
	}

	return nil
}

func resolveTSCCommand(entryPath, outDir string) (string, []string) {
	baseArgs := []string{
		"--target", "ESNext",
		"--module", "ESNext",
		"--moduleResolution", "node",
		"--lib", "ESNext,DOM",
		"--strict",
		"--skipLibCheck",
		"--pretty", "false",
		"--declaration", "false",
		"--sourceMap", "false",
		"--removeComments", "false",
		"--outDir", outDir,
		entryPath,
	}

	if _, err := exec.LookPath("tsc"); err == nil {
		return "tsc", baseArgs
	}

	return "npx", append([]string{"--yes", "tsc"}, baseArgs...)
}

func copyFile(srcPath, dstPath string) error {
	src, err := os.Open(srcPath)
	if err != nil {
		return err
	}
	defer src.Close()

	dst, err := os.Create(dstPath)
	if err != nil {
		return err
	}
	defer func() {
		_ = dst.Close()
	}()

	if _, err := io.Copy(dst, src); err != nil {
		return err
	}

	return dst.Close()
}

func runCleanup(cleanups []func()) {
	for i := len(cleanups) - 1; i >= 0; i-- {
		if cleanups[i] != nil {
			cleanups[i]()
		}
	}
}
